File: clients/ruby/model/worker_group.rb
module Model

# Group of workers on a planet
class WorkerGroup
    # Index of player controlling the workers
    attr_accessor :player_index
    # Number of workers in the group
    attr_accessor :number

    def initialize(player_index, number)
        @player_index = player_index
        @number = number
    end

    # Read WorkerGroup from input stream
    def self.read_from(stream)
        player_index = stream.read_int()
        number = stream.read_int()
        WorkerGroup.new(player_index, number)
    end

    # Write WorkerGroup to output stream
    def write_to(stream)
        stream.write_int(@player_index)
        stream.write_int(@number)
    end

    def to_s
        string_result = "WorkerGroup { "
        string_result += "player_index: "
        string_result += @player_index.to_s
        string_result += ", "
        string_result += "number: "
        string_result += @number.to_s
        string_result += " }"
        string_result
    end

    def to_str
        to_s
    end
end

end

File: clients/go/model/worker_group.go
package model

import "fmt"
import "io"
import . "spb_ai_champ/stream"

// Group of workers on a planet
type WorkerGroup struct {
    // Index of player controlling the workers
    PlayerIndex int32
    // Number of workers in the group
    Number int32
}

func NewWorkerGroup(playerIndex int32, number int32) WorkerGroup {
    return WorkerGroup {
        PlayerIndex: playerIndex,
        Number: number,
    }
}

// Read WorkerGroup from reader
func ReadWorkerGroup(reader io.Reader) WorkerGroup {
    var playerIndex int32
    playerIndex = ReadInt32(reader)
    var number int32
    number = ReadInt32(reader)
    return WorkerGroup {
        PlayerIndex: playerIndex,
        Number: number,
    }
}

// Write WorkerGroup to writer
func (workerGroup WorkerGroup) Write(writer io.Writer) {
    playerIndex := workerGroup.PlayerIndex
    WriteInt32(writer, playerIndex)
    number := workerGroup.Number
    WriteInt32(writer, number)
}

// Get string representation of WorkerGroup
func (workerGroup WorkerGroup) String() string {
    stringResult := "{ "
    stringResult += "PlayerIndex: "
    playerIndex := workerGroup.PlayerIndex
    stringResult += fmt.Sprint(playerIndex)
    stringResult += ", "
    stringResult += "Number: "
    number := workerGroup.Number
    stringResult += fmt.Sprint(number)
    stringResult += " }"
    return stringResult
}